### PR TITLE
Inject dev machine IP on Android and improve error message when connection fails

### PR DESCRIPTION
**Pull request by hrastnik**
## Summary:

I've implemented a feature that automatically bundles the Metro Bundler's IP address into Android builds. This change aligns the Android development experience with iOS, allowing the app to maintain a connection to the Metro Bundler even when disconnected from USB.

Currently, in iOS builds, the IP address of the computer running the Metro Bundler is automatically bundled into the app, ensuring seamless connectivity even when the device is disconnected from USB. In contrast, Android developers must manually input the IP address if the USB connection is lost, which can be tedious and error-prone.

I anticipate that a change where making IP the default method of connection will result in a lot of people running into issues where they can't connect to Metro server (for example, if they're on a different network, or they disable wifi). So I also changed the default error message you get in case the app can't connect to the bundler and updated the "Change Bundle Location" dev menu.

The previous error message

```
Unable to load script. Make sure you're either
running Metro (run 'npx react-native start') or
that your bundle 'RNTesterApp.android.bundle' is
packaged correctly for release.
```

was changed to:

```
Unable to load script.

Make sure you're running Metro (npx react-native start)
or that your bundle 'RNTesterApp.android.bundle' is
packaged correctly for release.

The device must be on the same WiFi as your laptop to
connect to Metro.

To use USB instead, shake the device to open the dev
menu and set the bundler location to 'localhost: 8081'
and run:
  adb reverse tcp:8081 tcp:8081

```

image

And the new dev menu UI looks like this:
image

The two buttons with "10.0.2.2:8081" and "localhost:8081" are suggestions which when tapped fill the input with the text from the button. The first button suggests the IP of the development machine, and the second one is hardcoded to localhost:8081.

## Changelog:

[ANDROID] [CHANGED] - Automatically use Metro bundler IP address when installing apps on Android

## Test Plan:

I've tested the implementation on a physical device and on emulator and it's working solid. However, I would invite further testing in order to catch possible edge cases.

### CODE REVIEW

File: packages/react-native/ReactAndroid/build.gradle.kts [hunk 1]
```diff
@@ -467,6 +469,14 @@ fun enableWarningsAsErrors(): Boolean {
   return value?.toString()?.toBoolean() ?: false
 }
 
+fun getHostIpAddress(): String =
+    NetworkInterface.getNetworkInterfaces().asSequence()
+        .filter { it.isUp && !it.isLoopback }
+        .flatMap { it.inetAddresses.asSequence() }
+        .filter { it is Inet4Address && !it.isLoopbackAddress }
+        .map { it.hostAddress }
+        .firstOrNull() ?: "127.0.0.1"
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/build.gradle.kts:478]: This won't work as it is. This IP address is the one used to build React Native (in this case it would be either your local machine or the CI machine that is publishing RN).

You set it as a buildConfigFlag so this value will be static and won't be re-accessed every time

**Review by cortinico — CHANGES_REQUESTED** [review]

File: packages/react-native/ReactAndroid/build.gradle.kts (hunk 1 above)
Comment (hrastnik) [reply packages/react-native/ReactAndroid/build.gradle.kts]: I've changed it to use resource value instead of BuildConfig. I think it should work now.

File: packages/gradle-plugin/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/AgpConfiguratorUtils.kt [hunk 2]
```diff
@@ -104,3 +107,11 @@ fun getPackageNameFromManifest(manifest: File): String? {
     return null
   }
 }
+
+fun getHostIpAddress(): String =
+  NetworkInterface.getNetworkInterfaces().asSequence()
+    .filter { it.isUp && !it.isLoopback }
+    .flatMap { it.inetAddresses.asSequence() }
+    .filter { it is Inet4Address && !it.isLoopbackAddress }
+    .map { it.hostAddress }
```
Comment (cortinico) [inline packages/gradle-plugin/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/AgpConfiguratorUtils.kt:116]: This seems like the correct approach. 

The only thing to be aware is that this should not prevent implementing Gradle configuration caching in the future: https://github.com/facebook/react-native/issues/45154

Comment (hrastnik) [reply packages/gradle-plugin/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/AgpConfiguratorUtils.kt:116]: @cortinico  I'm not sure I understand... Does this need additional changes or is it ok like this?

Comment (cortinico) [reply packages/gradle-plugin/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/AgpConfiguratorUtils.kt:116]: It requires extra caution (it's a reminder for myself) as this could prevent Gradle Configuration caching alltogether.

File: packages/gradle-plugin/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/AgpConfiguratorUtils.kt [hunk 3]
```diff
@@ -50,13 +52,14 @@ internal object AgpConfiguratorUtils {
     }
   }
 
-  fun configureDevPorts(project: Project) {
+  fun configureDevIpAndPorts(project: Project) {
```
Comment (huntie) [inline packages/gradle-plugin/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/AgpConfiguratorUtils.kt:55]: nit

```suggestion
  fun configureDevServerLocation(project: Project) {
```

**Review by huntie — COMMENTED** [review]
@hrastnik Thanks!

Would you be able to include a video recording in the test plan? Just so we can better understand the working flow :)

**Comment by hrastnik** [comment]
@huntie @cortinico I've recorded the most common scenarios and added to test plan. And I've merged the proposed wording changes.

File: packages/gradle-plugin/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/AgpConfiguratorUtils.kt [hunk 4]
```diff
@@ -104,3 +107,11 @@ fun getPackageNameFromManifest(manifest: File): String? {
     return null
   }
 }
+
+fun getHostIpAddress(): String =
```
Comment (cortinico) [inline packages/gradle-plugin/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/utils/AgpConfiguratorUtils.kt:111]:
```suggestion
internal fun getHostIpAddress(): String =
```

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.kt [hunk 5]
```diff
@@ -10,6 +10,7 @@ package com.facebook.react.modules.systeminfo
 import android.content.Context
 import android.os.Build
 import com.facebook.common.logging.FLog
+import com.facebook.react.BuildConfig
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.kt:13]: This is not needed right?

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.kt [hunk 6]
```diff
@@ -30,10 +31,10 @@ public object AndroidInfoHelpers {
   private fun isRunningOnStockEmulator(): Boolean =
       Build.FINGERPRINT.contains("generic") || Build.FINGERPRINT.startsWith("google/sdk_gphone")
 
-  @JvmStatic public fun getServerHost(port: Int): String = getServerIpAddress(port)
+  @JvmStatic public fun getServerHost(context: Context, port: Int): String = getServerIpAddress(context, port)
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.kt:34]: This is a breaking change. It has to be an overload (or a second parameter + `@JvmOverload`).

File: packages/react-native/ReactAndroid/src/main/jni/react/jni/JSLoader.cpp [hunk 7]
```diff
@@ -88,10 +88,13 @@ loadScriptFromAssets(AAssetManager* manager, const std::string& assetName) {
...
-      "either running Metro (run 'npx react-native start') or that your bundle '",
-      assetName,
-      "' is packaged correctly for release."));
+      "Unable to load script.\n\n"
+      "Make sure you're running Metro or that your "
+      "bundle '", assetName, "' is packaged correctly for release.\n\n"
+      "The device must be on the same Wi-Fi network as your laptop to connect to Metro.\n\n"
+      "To use USB instead, shake the device to open the Dev Menu and set "
+      "the bundler location to \"localhost:8081\" and run:\n"
+      "  adb reverse tcp:8081 tcp:8081"));
```
Comment (cortinico) [inline packages/react-native/ReactAndroid/src/main/jni/react/jni/JSLoader.cpp:97]: nice :)

**Review by huntie — APPROVED** [review]

**Review by cortinico — DISMISSED** [review]

File: packages/react-native/ReactAndroid/src/main/jni/react/jni/JSLoader.cpp (hunk 7 above)
Comment (huntie) [reply packages/react-native/ReactAndroid/src/main/jni/react/jni/JSLoader.cpp:97]: Ah, since we're here still, can we substitute "laptop" for "computer" too?

Also I'm not 100% on how line wrapping should work in Kotlin, but maybe these could be rebalanced? :)

Comment (cortinico) [reply packages/react-native/ReactAndroid/src/main/jni/react/jni/JSLoader.cpp:97]: @huntie when you import it, the formatter will take care of it

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.kt (hunk 5 above)
Comment (hrastnik) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.kt:13]: True

File: packages/react-native/ReactAndroid/src/main/jni/react/jni/JSLoader.cpp (hunk 7 above)
Comment (hrastnik) [reply packages/react-native/ReactAndroid/src/main/jni/react/jni/JSLoader.cpp]: I replaced "laptop" with "computer". Not sure about how to rebalance this better though.

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.kt (hunk 6 above)
Comment (hrastnik) [reply packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoHelpers.kt]: I returned support for `getServerHost(port: Int)` and in that case, when we have no context, I return `DEVICE_LOCALHOST` from `getServerIpAddress` (like in the previous implementation)

**Comment by cortinico** [comment]
This will have to be tested on top of:
- https://github.com/facebook/react-native/pull/49484

And making sure that running:

```
./gradlew build --configuration-cache
```

**twice** is not building anything. I suspect not, but let's do it before we merge it

**Comment by cortinico** [comment]
Hey @hrastnik 
I've tested this and it's not breaking Gradle Config Cache (which is good).

However, I'm a bit concerned about this as getting the machine IP by default AND using it as server address is a risky solution.

For example say I'm on a hotel wifi, and I can't find a way to connect the phone to the metro running on my laptop. And  I don't know if the hotel wifi is blocking TCP connections between hosts OR if I have the macos firewall enabled OR if my employer is blocking incoming TCP connections on my machine.

So we end up forcing users to 'debug' their network setup, just to avoid an `adb reverse` command which they're used to.

I believe this should be rolled out more gradually.
As a bare minimum:
1. Default to `localhost:8081` always, and let the user select the machine IP address we collected at build time if they wish.
2. Add a Gradle property to toggle if the build time IP should be used instead (that prop should be off by default)

**Comment by hrastnik** [comment]
@cortinico I understand the concern, but I still feel like using the IP as the server address should be the default.

1. It matches the iOS behaviour.
2. Most devs are developing RN apps either at home or work, with stable networks.
3. In case the network is unstable for any reason, and the app fails to connect to the Metro server, there are instruction on how to connect via USB.

However, if you feel like it's still too risky, I can update my PR and add a gradle property that would control the behaviour.

**Comment by hrastnik** [comment]
@cortinico Any updates on this? Should I add the Gradle property, or is there any chance we can roll with it as is?

**Comment by cortinico** [comment]
> @cortinico Any updates on this? Should I add the Gradle property, or is there any chance we can roll with it as is?

I personally tested this on both our corp wifi (at Meta) and in a hotel wifi when I was travelling, and neither scenarios worked out of the box. 

I'm really concerned that if we ship this as it is by default, we're going to create more frustration than benefits

I'd suggest we make it optional and potentially change it to be a default in a later release

**Comment by cortinico** [comment]
> I'd suggest we make it optional and potentially change it to be a default in a later release

@hrastnik are you looking into this one or do you need any support?

**Comment by hrastnik** [comment]
@cortinico I've been swamped the last few days so any support would be much appreciated.

**Comment by cortinico** [comment]
> @cortinico I've been swamped the last few days so any support would be much appreciated.

I'll try to see if I can find some time on this over the next week or two

**Comment by cortinico** [comment]
I've spent some time on this as I had to refactor your code (moslty because `DevSupportManagerBase` is now being migrated to Kotlin + it was already a 1K LOC file).

We're going to go with a first iteration where the new bundle selection dialog is shipped and users can pick either localhost or the machine IP. 

In the next iteration we're going to change this to default to the machine IP

**Comment by hrastnik** [comment]
Sounds good! Makes sense.
